### PR TITLE
Documentation HowStuffWorks.md

### PR DESCRIPTION
**Pull request by barrystaes**
For grasping the overall design i just wrote [HowStuffWorks.md](https://github.com/barrystaes/react-starter/blob/documentation-howstuffworks/NOTES/HowStuffWorks.md) with some nudges in the right direction. No details by intention.
I would like to add `/README.md` is of great quality as-is, it got me up and running in notime.

I think its about time to reconcile this with **react-starter**, so more people can contribute.
For example, i am still sorting out how the stores work: especially that section needs more attention. Any information on this is welcome! I shall amend the document while the PR is pending.

Related to https://github.com/webpack/react-starter/issues/17 and https://github.com/webpack/react-starter/pull/49

### CODE REVIEW

File: NOTES/HowStuffWorks.md [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+### production mode
+In this mode, the React HTML output is (pre)rendered (and populated) on the server (a.k.a. isomorphic).
+
+Run the server using `npm run start`.
+It wil use `/lib/server-production.js` which wil use `/lib/server.js` which wil use `/config/prerender.jsx` which wil use `/app/prerender.html`. A big HTML file is loaded (and output is shown immediately), the JS is downloaded + executed + rendered, and output is updated.
+
+This server side render (prerender) is possible because the React JS can be executed on the server.
+In your browser the main `React.render(<Application />, document.getElementById("content"))` call (in `/config/app.jsx`) outputs to the browser DOM.
+But when prerendering, the main `React.renderToString(<Application />)` call (in `/app/prerender.jsx`) outputs to a string, which is inserted into the HTML (including React component states) as content.
```
Comment (sokra) [inline NOTES/HowStuffWorks.md]: You should tell about the fact that the application is compiled with webpack for prerendering using `node` target. This way JSX can be used and there is no trouble when requiring stylesheets and other static assets.

Comment (barrystaes) [inline NOTES/HowStuffWorks.md]: Good one thanks. I'll make a "How the build works" with webpack train-of-thought. I'm going to include some questions to you, parts are not entirely clear to me yet.

Comment (sokra) [inline NOTES/HowStuffWorks.md]: Ok. Feel free to ask.

Comment (barrystaes) [inline NOTES/HowStuffWorks.md]: I was not aware my comment was already visible.

**Comment by barrystaes** [comment]
FYI i wont be able to respond/work on this for a week or so.

File: NOTES/HowStuffWorks.md (hunk 1 above)
Comment (sokra) [inline NOTES/HowStuffWorks.md]: `webpack.config.js` is not unused. It is used when you just invoke the `webpack` command. It is there to make custom builds from command line. I. e. when you want to experiment with something...

Comment (barrystaes) [inline NOTES/HowStuffWorks.md]: Ok that should be added to HowStuffWords.md build section then.

File: NOTES/HowStuffWorks.md [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+- sets up a queue that only allows one REST request at a time, and aggregates subsequent changes. 
+  These are then sent as one request.. :question:
+- and ultimately constructs the two respective `ItemStore` objects (based on `/app/mainStoresDescriptions.js`) to which it assigns these routines, queue, and the initial data. This initial data may have been inserted when `/app/prerender.html`.
+- These are then exported back to `/config/app.jsx`, along with a store named Router. This store just holds a `transition` value which is later read in `/Application/index.jsx` to show "loading...' in UI.
+
+### Stores in use
+
+todo :question: this section needs review and should be extended.
+
+Note on `Application.update()`: normally components should not access the stores except for reading in `getState()`, as demonstrated with the Router store. Everything else should be done with actions.
```
Comment (barrystaes) [inline NOTES/HowStuffWorks.md]: @sokra i feel that you'll probably have something to add on the `How the stores work` section above. You grasp that way better than me;

- How is it triggered to refresh everything from the server,
- how does it optimize this diff when loading from server - if at all,
- how does it propagate changes when the user edits items, and
- how do values travel before ending up in some components `render()` ?
- how does the queue combine multiple requests? I cant imagine subsequent add/remove/edits would result in just one rest call.. do they?

Amongst other paths that the store data travels that you may know of that i dont..

Comment (barrystaes) [inline NOTES/HowStuffWorks.md]: @sokra i feel that you'll probably have some remarks on my initial `How the stores work` section above. You grasp that way better than me;

- How is it triggered to refresh everything from the server,
- how does it optimize this diff when loading from server - if at all,
- how does it propagate changes when the user edits items, and
- how do values travel before ending up in some components `render()` ?
- how does the queue combine multiple requests? I cant imagine subsequent add/remove/edits would result in just one rest call.. do they?

Amongst other paths that the store data travels that you may know of that i dont..

Comment (sokra) [inline NOTES/HowStuffWorks.md]: > How is it triggered to refresh everything from the server,

`config/app.jsx` invalidates store data when you change the page. When the pages read the data again it is invalid and will be refetched.

> how does it propagate changes when the user edits items, and

The component fires an action (from `app/actions.jsx`) on edit. The action is handled in `app/mainStores.jsx` and writes some items on stores. The stores update their internal cache (which is displayed on the page as optimistic update) and do server requests (through their registered handlers in `app/mainStores.jsx`.

> how do values travel before ending up in some components render() ?

component requests values from store in `getState()` -> `app/mainStores.jsx` read (unavailable or invalidated) data from server -> internal items-store cache for the store -> `getState()` -> components `this.state` -> `render()`

> how does the queue combine multiple requests? I cant imagine subsequent add/remove/edits would result in just one rest call.. do they?

items-store store a single `update` per entry. Any subsequent `updateItem()` call causes both updates to be merged (`mergeUpdates` from items-store). Requests from stores to the server are queued (`queueRequest` in `app/mainStores.jsx`). Here only a single ongoing request is allowed. All writes that happen in the meantime are merged into a single write.

**Comment by barrystaes** [comment]
:information_source: Note: current document state is interim / not fit for merge. Comments welcome though.

**Comment by barrystaes** [comment]
At this stage, i dont see how i can improve this document except for adding (to much) paragraphs to explain details on parts that are way better explained in their respective documentation.

I think its best to have this merged and have it evolve from there:
- A few parts (marked with :question:) need insights that i dont have.
- A quick writeup on the *How the webpack configuration works* would be nice to complement the *How the build works* section.
- Remember to make sure that any section title is an actual question that developers ask themselves when taking a first look at this project.

**Pull request merged**